### DISCUSSION TITLE
How to map the atoms after RunReactants?

### DISCUSSION THREAD

**Discussion by IannLiu** [rdkit/rdkit#4327] [Q&A]
After RunReactants, how to map from the atoms of the products to the atoms of the reactants?
The reactant SMILES and reaction template:
`reactants = '[H:6][C:1]([H:7])([H:8])[H:2].[C:3]([C:4]([H:9])([H:10])[Cl:5])([H:11])([H:12])[H:13]'`
`react_temp = '[C:1]-[H:2].[C:3]-[Cl:4]>>[C:1]-[Cl:4].[C:3]-[H:2]'`
Here is my code:
`RDKIT_SMILES_PARSER_PARAMS = Chem.SmilesParserParams()`
`def str_to_mol(string: str, explicit_hydrogens: bool = True) -> Chem.Mol:`

    if string.startswith('InChI'):
        mol = Chem.MolFromInchi(string, removeHs=not explicit_hydrogens)
    else:
        RDKIT_SMILES_PARSER_PARAMS.removeHs = not explicit_hydrogens
        mol = Chem.MolFromSmiles(string, RDKIT_SMILES_PARSER_PARAMS)
    if explicit_hydrogens:
        return mol
    else:
        return Chem.RemoveHs(mol)

`reactants_smiles = reactants.split('.')`
`mol1 = str_to_mol(reactants_smiles[0])`
`mol2 = str_to_mol(reactants_smiles[1])`
`outcomes_rdkit_mol = AllChem.ReactionFromSmarts(react_temp).RunReactants((mol1,mol2))`

`if outcomes_rdkit_mol:`

    print("outcome generated")
    for outcome in outcomes_rdkit_mol:
        for m in outcome:
            print(Chem.MolToSmiles(m))
            for a in m.GetAtoms():
                if a.HasProp('old_mapno'):
                    print('the changed atom symbol is {},atom index in reactants is {}'. format(a.GetSymbol(), a.GetProp('react_atom_idx')))
        print("************************")
else:
    print("failed")`

And the output:
outcome generated
ClC([H:2])([H:7])[H:8]
the changed atom symbol is C, atom index in reactants is 1
the changed atom symbol is Cl, atom index in reactants is 4
[H]C([C:3]([H:11])([H:12])[H:13])([H:9])[H:10]
the changed atom symbol is C,  atom index in reactants is 1
the changed atom symbol is H, atom index in reactants is 0
************************
ClC([H:2])([H:6])[H:8]
the changed atom symbol is C, atom index in reactants is 1
the changed atom symbol is Cl, atom index in reactants is 4
[H]C([C:3]([H:11])([H:12])[H:13])([H:9])[H:10]
the changed atom symbol is C, atom index in reactants is 1
the changed atom symbol is H, atom index in reactants is 2
************************
ClC([H:2])([H:6])[H:7]
the changed atom symbol is C, atom index in reactants is 1
the changed atom symbol is Cl, atom index in reactants is 4
[H]C([C:3]([H:11])([H:12])[H:13])([H:9])[H:10]
the changed atom symbol is C, atom index in reactants is 1
the changed atom symbol is H, atom index in reactants is 3
************************
ClC([H:6])([H:7])[H:8]
the changed atom symbol is C, atom index in reactants is 1
the changed atom symbol is Cl, atom index in reactants is 4
[H]C([C:3]([H:11])([H:12])[H:13])([H:9])[H:10]
the changed atom symbol is C, atom index in reactants is 1
the changed atom symbol is H, atom index in reactants is 4
************************
There are two different reactants. How can I know which reactant the atomic index(obtained from outcome_atom.GetProp('react_atom_idx') ) belongs to?

**Comment by greglandrum** [comment]
@YanLiuML: There are two parts to this.
Here's a bit of code to demonstrate what you need:
```
>>> rxn = rdChemReactions.ReactionFromSmarts('[C:1]-[O:2].[C:3](=[O:4])[OH]>>[C:1]-[O:2]-[C:3]=[O:4]')
>>> reactants = [Chem.MolFromSmiles(x) for x in ('CCO','OC=O')]
>>> for i,m in enumerate(reactants):
...     for atom in m.GetAtoms():
...         atom.SetIntProp('reactant_idx',i)
... 
>>> ps = rxn.RunReactants(reactants)
>>> p0 = ps[0][0]
>>> for atom in p0.GetAtoms():
...     print(atom.GetIdx(),atom.GetPropsAsDict())
... 
0 {'old_mapno': 1, 'react_atom_idx': 1}
1 {'_ReactionDegreeChanged': 1, 'old_mapno': 2, 'react_atom_idx': 2}
2 {'old_mapno': 3, 'react_atom_idx': 1}
3 {'old_mapno': 4, 'react_atom_idx': 2}
4 {'__computedProps': <rdkit.rdBase._vectNSt7__cxx1112basic_stringIcSt11char_traitsIcESaIcEEE object at 0x7f9ac5161270>, '_CIPRank': 0, 'reactant_idx': 0, 'react_atom_idx': 0}
>>> atomMapToReactantMap={}
>>> for ri in range(rxn.GetNumReactantTemplates()):
...     rt = rxn.GetReactantTemplate(ri)
...     for atom in rt.GetAtoms():
...         if atom.GetAtomMapNum():
...             atomMapToReactantMap[atom.GetAtomMapNum()] = ri
... 
>>> 
>>> atomMapToReactantMap
{1: 0, 2: 0, 3: 1, 4: 1}
```
Atoms which are just copied over directly from the reactants (those which were not mapped in the reaction) will still have the `reactant_idx` property set.
The reacting atoms will have the `old_mapno` property set, and you can lookup which reactant those came from in the `atomMapToReactantMap` dictionary.

**Reply by IannLiu** [reply]
Thanks for your reply.  The following steps can be done:
Firstly, mapping {reactant flag -> atom map number of reactant template}
mapping{atom map number of reacting atoms of products -> atom map number of template}
Then, traceback: reacting atoms ->  which reactant belongs to?
Finally,  get react_atom_index of reacting atoms, and set reacting atoms' atom map number from
Reactant numbers.

There is another question related to RunReactant. I want to generate products from reactants(explicit hydrogen with atom map for all atoms). Here are the reactants:
`reactants = '[H:6][C:1]([H:7])([H:8])[Br:2].[C:3]([C:4]([H:9])([H:10])[Cl:5])([H:11])([H:12])[H:13]'`
Here are some templates I tested:
`react_temp1 = '[C:1]-[Br:2].[C:3]-[Cl:4]>>[C:1]-[Cl:4].[C:3]-[Br:2] #True, 1 outcome`
`react_temp2 = '[Br;D1;+0:1]-[C;D4;+0:2].[C;D4;+0:7](-[Cl;D1;+0:8])>>[Br;D1;+0:1]-[C;D4;+0:7].[Cl;D1;+0:8]-[C;D4;+0:2]' #True, 1 outcome`
`react_temp3 = '[Br;H0;D1;+0:1]-[C;H3;D4;+0:2].[C:6]-[C;H2;D4;+0:7](-[Cl;H0;D1;+0:8])>>[Br;H0;D1;+0:1]-[C;H2;D4;+0:7](-[C:6]).[Cl;H0;D1;+0:8]-[C;H3;D4;+0:2]'#True, 1 outcome`
`react_temp4 = '[Br;D1;+0:1]-[C;D4;+0:2](-[H;D1:3])(-[H;D1:4])-[H;D1:5].[C:6]-[C;D4;+0:7](-[Cl;D1;+0:8])(-[H;D1:9])-[H;D1:10]>>[Br;D1;+0:1]-[C;D4;+0:7](-[C:6])(-[H;D1:9])-[H;D1:10].[Cl;D1;+0:8]-[C;D4;+0:2](-[H;D1:3])(-[H;D1:4])-[H;D1:5]'#both D<n> and X<n> are tested, False, 0 outcome`
`react_temp5 = '[Br;H0;D1;+0:1]-[C;H0;D4;+0:2](-[H;D1;H0:3])(-[H;D1;H0:4])-[H;D1;H0:5].[C:6]-[C;H0;D4;+0:7](-[Cl;H0;D1;+0:8])(-[H;D1;H0:9])-[H;D1;H0:10]>>[Br;H0;D1;+0:1]-[C;H0;D4;+0:7](-[C:6])(-[H;D1;H0:9])-[H;D1;H0:10].[Cl;H0;D1;+0:8]-[C;H0;D4;+0:2](-[H;D1;H0:3])(-[H;D1;H0:4])-[H;D1;H0:5]' #False, 0 outcome`
`react_temp6 = '[Br;D1;+0:1]-[C;D4;+0:2](-[#1;D1:3])(-[#1;D1:4])-[#1;D1:5].[C:6]-[C;D4;+0:7](-[Cl;D1;+0:8])(-[#1;D1:9])-[#1;D1:10]>>[Br;D1;+0:1]-[C;D4;+0:7](-[C:6])(-[#1;D1:9])-[#1;D1:10].[Cl;D1;+0:8]-[C;D4;+0:2](-[#1;D1:3])(-[#1;D1:4])-[#1;D1:5]' # 12 outcomes, with error:`
>>>RDKit ERROR: getExplicitValence() called without call to calcExplicitValence()
RDKit ERROR: Violation occurred on line 199 in file C:\Users\glandrum\Anaconda3\conda- bld\rdkit_1603168709418\work\Code\GraphMol\Atom.cpp
RDKit ERROR: Failed Expression: d_explicitValence > -1
RDKit ERROR: ****
RDKit ERROR: 
RDKit ERROR: [22:11:39] 
RDKit ERROR: 
RDKit ERROR: ****
RDKit ERROR: Pre-condition Violation
RDKit ERROR: getExplicitValence() called without call to calcExplicitValence()

Here are some questions:
1. Why can't use the strict SMARTS symbol? (For example:  react_temp5)
2. For the react_temp4 and react_temp6, the difference is: symbol 'H' is replaced by '#1'. Why does react_temp6 work?
3. Why did something go wrong when I ran the react_temp6?
4. Although react_temp3 generates one outcome, the SMILES symbol is wrong when I employ outcome.MolToSmiles. The results is `'[Br:2][CH2:4]([C:3]([H:11])([H:12])[H:13])([H:9])[H:10]'`
4. Is there anything I should pay attention to when generating templates( reactants with explicit hydrogen, and with atom map for all atoms)?

**Reply by greglandrum** [reply]
It's impossible to answer detailed questions like this unless you provide a code sample we can run which demonstrates the problem you're having.

**Reply by IannLiu** [reply]
Thanks for your reply. I'm sorry to forget to provide my code. Here is my code:
The reactant smiles:
`reactants = '[H:6][C:1]([H:7])([H:8])[Br:2].[C:3]([C:4]([H:9])([H:10])[Cl:5])([H:11])([H:12])[H:13]'`
We expect to obtain the products:
`prods = '[H:6][C:1]([H:7])([H:8])[Cl:5].[C:3]([C:4]([H:9])([H:10])[Br:2])([H:11])([H:12])[H:13]'`
Here are some templates:
`react_temp1 = '[C:1]-[Br:2].[C:3]-[Cl:4]>>[C:1]-[Cl:4].[C:3]-[Br:2]`
`react_temp2 = '[Br;D1;+0:1]-[C;D4;+0:2].[C;D4;+0:7](-[Cl;D1;+0:8])>>[Br;D1;+0:1]-[C;D4;+0:7].[Cl;D1;+0:8]-[C;D4;+0:2]'`
`react_temp3 = '[Br;H0;D1;+0:1]-[C;H3;D4;+0:2].[C:6]-[C;H2;D4;+0:7](-[Cl;H0;D1;+0:8])>>[Br;H0;D1;+0:1]-[C;H2;D4;+0:7](-[C:6]).[Cl;H0;D1;+0:8]-[C;H3;D4;+0:2]'`
`react_temp4 = '[Br;D1;+0:1]-[C;D4;+0:2](-[H;D1:3])(-[H;D1:4])-[H;D1:5].[C:6]-[C;D4;+0:7](-[Cl;D1;+0:8])(-[H;D1:9])-[H;D1:10]>>[Br;D1;+0:1]-[C;D4;+0:7](-[C:6])(-[H;D1:9])-[H;D1:10].[Cl;D1;+0:8]-[C;D4;+0:2](-[H;D1:3])(-[H;D1:4])-[H;D1:5]'`
`react_temp5 = '[Br;H0;D1;+0:1]-[C;H0;D4;+0:2](-[H;D1;H0:3])(-[H;D1;H0:4])-[H;D1;H0:5].[C:6]-[C;H0;D4;+0:7](-[Cl;H0;D1;+0:8])(-[H;D1;H0:9])-[H;D1;H0:10]>>[Br;H0;D1;+0:1]-[C;H0;D4;+0:7](-[C:6])(-[H;D1;H0:9])-[H;D1;H0:10].[Cl;H0;D1;+0:8]-[C;H0;D4;+0:2](-[H;D1;H0:3])(-[H;D1;H0:4])-[H;D1;H0:5]'`
`react_temp6 = '[Br;D1;+0:1]-[C;D4;+0:2](-[#1;D1:3])(-[#1;D1:4])-[#1;D1:5].[C:6]-[C;D4;+0:7](-[Cl;D1;+0:8])(-[#1;D1:9])-[#1;D1:10]>>[Br;D1;+0:1]-[C;D4;+0:7](-[C:6])(-[#1;D1:9])-[#1;D1:10].[Cl;D1;+0:8]-[C;D4;+0:2](-[#1;D1:3])(-[#1;D1:4])-[#1;D1:5]'`

Then, I defined a function to conver the SMILES to mol type:
`def str_to_mol(string: str, explicit_hydrogens: bool = True) -> Chem.Mol:`

    if string.startswith('InChI'):
        mol = Chem.MolFromInchi(string, removeHs=not explicit_hydrogens)
    else:
        RDKIT_SMILES_PARSER_PARAMS.removeHs = not explicit_hydrogens
        mol = Chem.MolFromSmiles(string, RDKIT_SMILES_PARSER_PARAMS)
    if explicit_hydrogens:
        return mol
    else:
        return Chem.RemoveHs(mol)

Next, initializing the reactants. For atoms of templates, mapping which reactant belongs to:
`reactants_smiles = reactants.split('.')`
`reacts_mol = [str_to_mol(r) for r in reactants_smiles]`
`rxn = AllChem.ReactionFromSmarts(react_temp)`
`display(rxn)`

`atomMapToReactantMap={}`
`for ri in range(rxn.GetNumReactantTemplates()):`

    rt = rxn.GetReactantTemplate(ri)
    for atom in rt.GetAtoms():
        atomMapToReactantMap[atom.GetAtomMapNum()] = ri

We run the reactants and template:
`outcomes_rdkit_mol = rxn.RunReactants(reacts_mol)`

The following part is unnecessary, it resets the atom map number of products so that the atoms of products can be traced back.
`if outcomes_rdkit_mol:`

    for outcome in outcomes_rdkit_mol:
        for i, m in enumerate(outcome):
            display(m)
            for a in m.GetAtoms():
                if a.HasProp('old_mapno'):
                    r_num = atomMapToReactantMap[int(a.GetProp('old_mapno'))]
                    react_idx = int(a.GetProp('react_atom_idx'))
                    react_num =reacts_mol[r_num].GetAtomWithIdx(react_idx).GetAtomMapNum()
                    a.SetAtomMapNum(react_num)
            Chem.SanitizeMol(m)
            print('Outcome smiles: ', Chem.MolToSmiles(m))
            if i == 0:
                prod = m
            else:
                prods = AllChem.CombineMols(prod, m)
        display(prods)

Then, we run different templates in order.
For react_temp1 and react_template2, 1 outcome(with correct SMILES) was obtained.
For react_temp3, 1 outcome was obtained, but the SMILES of products is wrong`'[CH3:1]([Cl:5])([H:6])([H:7])[H:8]'`. 
We generated a strict template(react_temp5) which includes the properties: atom symbol, degree, charge, hydrogen numbers, atom map, but there is no outcome.
 Then, we remove the hydrogen numbers and obtained react_temp4, but there is no outcome.
Then, we change the 'H' symbol of react_temp4 to '#1', which is the react_temp6.  12 outcomes were obtained.

For another reaction:
`reactants1 = '[H:6][C:1]([H:7])([H:8])[H:2].[C:3]([C:4]([H:9])([H:10])[Cl:5])([H:11])([H:12])[H:13]'`
We expect to obtain the products:
`prods1 ='[H:6][C:1]([H:7])([H:8])[Cl:5].[C:3]([C:4]([H:9])([H:10])[H:2])([H:11])([H:12])[H:13]'`
The reaction template is:
`react_temp7 = '[CH4;D4;+0:1].[C:2]-[CH2;D4;+0:3]-[Cl;H0;D1;+0:4]>>[CH3;D4;+0:1]-[Cl;H0;D1;+0:4].[C:2]-[CH3;D4;+0:3]'`
We obtained 1 outcome, but the products are wrong as well.
It seems that reaction template must be 'explicit hydorgen' if the reactants are 'explicit hydorgen'.
Therefore, we replaced react_temp7 by:
`react_temp8 = '[#1;D1;+0:1]-[C;D4;+0:2](-[#1;D1:3])(-[#1;D1:4])-[#1;D1:5].[C:6]-[C;D4;+0:7](-[Cl;D1;+0:8])(-[#1;D1:9])-[#1;D1:10]>>[#1;D1;+0:1]-[C;D4;+0:7](-[C:6])(-[#1;D1:9])-[#1;D1:10].[Cl;D1;+0:8]-[C;D4;+0:2](-[#1;D1:3])(-[#1;D1:4])-[#1;D1:5]`
and 48 outcomes are obtained. We also obtained correct SMILES.
However, for the free radical reaction:
`reactants3 = '[H:6][C:1]([H:7])([H:8]).[C:3]([C:4]([H:9])([H:10])[Cl:5])([H:11])([H:12])[H:13]'`
We expect to obtain the products:
`[C:1]-[C;D4;+0:2](-[Cl;D1;+0:3])(-[H;D1:4])-[H;D1:5].[H;D1:6]-[C;D3;+0:7](-[H;D1:8])-[H;D1:9]>>[C:1]-[C;D3;+0:2](-[H;D1:4])-[H;D1:5].[Cl;D1;+0:3]-[C;D4;+0:7](-[H;D1:6])(-[H;D1:8])-[H;D1:9]`
and template is:
`react_temp9 = '[#1;D1:6]-[C;D3;+0:7](-[#1;D1:8])-[#1;D1:9].[C:1]-[C;D4;+0:2](-[Cl;D1;+0:3])(-[#1;D1:4])-[#1;D1:5]>>[C:1]-[C;D3;+0:2](-[#1;D1:4])-[#1;D1:5].[Cl;D1;+0:3]-[C;D4;+0:7](-[#1;D1:6])(-[#1;D1:8])-[#1;D1:9]'`
The SMILES of the product(with free radicals) is:
`[C:3]([CH:4]([H:9])[H:10])([H:11])([H:12])[H:13]`
One more hydrogen atom was added to the carbon atom 4.

These days, we are generating reaction templates and calculating the outcomes for radical reactions. For example, hydrogen abstraction. Therefore, explicit hydrogen is necessary. Here are the questions I want to ask:
1. Why can't use the strict SMARTS symbol? (For example: react_temp5, none outcomes)
2. For the react_temp4 and react_temp6, the difference is: symbol 'H' is replaced by '#1'. Why does react_temp4 can not work?
3. How to obtain correct SMILES for free radical reactions(react_temp9)?
4. Is there anything I should pay attention to when generating templates for free radical reactions( reactants with explicit hydrogen, and with atom map for all atoms)?

**Reply by greglandrum** [reply]
Thanks for providing more information, but what I really need to be able to figure out what's going on and help here is a code snippet demonstrating the problem that I can copy and paste into a file and run it. I'd like to help, but I can't spend an hour interpreting what you did and trying to create a reproducible so that I can actually see the problem and suggest a solution.

**Reply by IannLiu** [reply]
Thanks for your reply.  I want to generate products of free radical reactions using reaction templates and reactants. Both  templates and reactants have explicit hydrogens.  Here is a case:
The reactants:
`reactants = '[H:6][C:1]([H:7])([H:8]).[C:3]([C:4]([H:9])([H:10])[Cl:5])([H:11])([H:12])[H:13]'`
The template:
`react_temp = '[#1;D1:6]-[C;D3;+0:7](-[#1;D1:8])-[#1;D1:9].[C:1]-[C;D4;+0:2](-[Cl;D1;+0:3])(-[#1;D1:4])-[#1;D1:5]>>[C:1]-[C;D3;+0:2](-[#1;D1:4])-[#1;D1:5].[Cl;D1;+0:3]-[C;D4;+0:7](-[#1;D1:6])(-[#1;D1:8])-[#1;D1:9]'`
Then, I defined a function to conver the SMILES to mol type:
`def str_to_mol(string: str, explicit_hydrogens: bool = True) -> Chem.Mol:`

    if string.startswith('InChI'):
        mol = Chem.MolFromInchi(string, removeHs=not explicit_hydrogens)
    else:
        RDKIT_SMILES_PARSER_PARAMS.removeHs = not explicit_hydrogens
        mol = Chem.MolFromSmiles(string, RDKIT_SMILES_PARSER_PARAMS)
    if explicit_hydrogens:
        return mol
    else:
        return Chem.RemoveHs(mol)
Then, we run the reactant:
`rxn = AllChem.ReactionFromSmarts(react_temp)`
`outcomes_rdkit_mol = rxn.RunReactants([str_to_mol(reactant) for reactant in reactants.split('.')])`
However, the free radical of the product disappears, and a hydrogen atom is automaticly addadded. We got two molecules with no radical.  However, in product template, we defined the free radical fragment: `[C:1]-[C;D3;+0:2](-[#1;D1:4])-[#1;D1:5]`. 
Is there something wrong with the template I defined?
